apps/emqx_gateway/src/emqx_gateway_cm_registry.erl

```diff
@@ -0,0 +1,140 @@
+
+%% @doc Register a global channel.
+-spec register_channel(atom(), binary() | {binary(), pid()}) -> ok.
+register_channel(Type, ClientId) when is_binary(ClientId) ->
+    register_channel(Type, {ClientId, self()});
+
+register_channel(Type, {ClientId, ChanPid}) when is_binary(ClientId), is_pid(ChanPid) ->
+    ekka_mnesia:dirty_write(tabname(Type), record(ClientId, ChanPid)).
+
+%% @doc Unregister a global channel.
+-spec unregister_channel(atom(), binary() | {binary(), pid()}) -> ok.
+unregister_channel(Type, ClientId) when is_binary(ClientId) ->
+    unregister_channel(Type, {ClientId, self()});
+
+unregister_channel(Type, {ClientId, ChanPid}) when is_binary(ClientId), is_pid(ChanPid) ->
+    ekka_mnesia:dirty_delete_object(tabname(Type), record(ClientId, ChanPid)).
+
+%% @doc Lookup the global channels.
+-spec lookup_channels(atom(), binary()) -> list(pid()).
+lookup_channels(Type, ClientId) ->
+    [ChanPid || #channel{pid = ChanPid} <- mnesia:dirty_read(tabname(Type), ClientId)].
+
+record(ClientId, ChanPid) ->
+    #channel{chid = ClientId, pid = ChanPid}.
+
+%%--------------------------------------------------------------------
+%% gen_server callbacks
+%%--------------------------------------------------------------------
+
+init([Type]) ->
+    Tab = tabname(Type),
+    ok = ekka_mnesia:create_table(Tab, [
+                {type, bag},
+                {ram_copies, [node()]},
+                {record_name, channel},
+                {attributes, record_info(fields, channel)},
+                {storage_properties, [{ets, [{read_concurrency, true},
+                                             {write_concurrency, true}]}]}]),
+    ok = ekka_mnesia:copy_table(Tab, ram_copies),
+    %%ok = ekka_rlog:wait_for_shards([?CM_SHARD], infinity),
+    ok = ekka:monitor(membership),
+    {ok, #{type => Type}}.
+
+handle_call(Req, _From, State) ->
+    logger:error("Unexpected call: ~p", [Req]),
+    {reply, ignored, State}.
+
+handle_cast(Msg, State) ->
+    logger:error("Unexpected cast: ~p", [Msg]),
+    {noreply, State}.
+
+handle_info({membership, {mnesia, down, Node}}, State = #{type := Type}) ->
+    Tab = tabname(Type),
+    global:trans({?LOCK, self()},
+                 fun() ->
+                     ekka_mnesia:transaction(fun cleanup_channels/2, [Node, Tab])
+                 end),
+    {noreply, State};
+
+handle_info({membership, _Event}, State) ->
+    {noreply, State};
+
+handle_info(Info, State) ->
+    logger:error("Unexpected info: ~p", [Info]),
+    {noreply, State}.
+
+terminate(_Reason, _State) ->
+    ok.
+
+code_change(_OldVsn, State, _Extra) ->
+    {ok, State}.
+
+%%--------------------------------------------------------------------
+%% Internal functions
+%%--------------------------------------------------------------------
+
+cleanup_channels(Node, Tab) ->
+    Pat = [{#channel{pid = '$1', _ = '_'}, [{'==', {node, '$1'}, Node}], ['$_']}],
+    lists:foreach(fun(Chan) ->
+        mnesia:delete_object(Tab, Chan, write)
```

Comment:
ekka_mnesia

Reply:
Do I need to replace `mnesia:select` as well?

Reply:
emmm..I just checked that ekka_mnesia does not implement `delete_object/3` (ekka-v0.10.2)

Reply:
Sorry, my bad! `delete_object` is fine, I read it as "dirty_delete_object" :facepalm:

Reply:
Lol~ ok